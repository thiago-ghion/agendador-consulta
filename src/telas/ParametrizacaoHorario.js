import React, { useEffect, useState, useRef } from "react";
import { useSelector, useDispatch } from "react-redux";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Container from "react-bootstrap/Container";
import Form from "react-bootstrap/Form";
import { Button } from "react-bootstrap";
import ToolkitProvider, {
  Search,
} from "react-bootstrap-table2-toolkit/dist/react-bootstrap-table2-toolkit.min";
import BootstrapTable from "react-bootstrap-table-next";
import paginationFactory from "react-bootstrap-table2-paginator";

import BotaoExportar from "../componentes/BotaoExportar";
import {
  listarHorarioAction,
  registrarHorarioAction,
  ativarHorarioAction,
  desativarHorarioAction,
} from "../features/horarioSlice";

import "react-bootstrap-table-next/dist/react-bootstrap-table2.css";
import "react-bootstrap-table2-paginator/dist/react-bootstrap-table2-paginator.min.css";
import "react-bootstrap-table2-toolkit/dist/react-bootstrap-table2-toolkit.min.css";

function ParametrizacaoHorario() {
  const dispatch = useDispatch();

  const { lista } = useSelector((state) => state.horario);
  const [horario, setHorario] = useState("");
  let horarioInput;

  const telaCarregada = useRef(false);

  useEffect(() => {
    if (telaCarregada.current === false) {
      const prepararLista = async () => {
        await dispatch(listarHorarioAction());
      };
      prepararLista();
      telaCarregada.current = true;
    }
  }, [dispatch]);

  const registrar = async () => {
    const resposta = await dispatch(
      registrarHorarioAction({ idHorario: horario })
    );
    if (resposta.error === undefined) {
      setHorario("");
      dispatch(listarHorarioAction());
    } else {
      horarioInput.focus();
    }
  };

  function listaFormatter(cell, row, rowIndex, formatExtraData) {
    return (
      <Button
        variant={row.indicadorAtivo === "S" ? "danger" : "success"}
        onClick={() => {
          if (row.indicadorAtivo === "S") {
            dispatch(desativarHorarioAction({ idHorario: row.idHorario }));
          }

          if (row.indicadorAtivo === "N") {
            dispatch(ativarHorarioAction({ idHorario: row.idHorario }));
          }
        }}
      >
        {row.indicadorAtivo === "S"
          ? "Desativar"
          : "\u00a0\u00a0\u00a0Ativar\u00a0\u00a0\u00a0"}
      </Button>
    );
  }

  const colunas = [
    {
      dataField: "idHorario",
      text: "#",
      sort: true,
      style: { cursor: "pointer" },
      headerStyle: () => {
        return { width: "80px" };
      },
    },
    {
      dataField: "horario",
      text: "Horário",
      style: { cursor: "pointer" },
      sort: true,
    },
    {
      dataField: "indicadorAtivo",
      text: "Ativação",
      style: { cursor: "pointer" },
      sort: true,
      formatter: listaFormatter,
    },
  ];

  const defaultSorted = [
    {
      dataField: "idHorario",
      order: "asc",
    },
  ];

  const pagination = paginationFactory({
    page: 1,
    sizePerPage: 10,
    lastPageText: ">>",
    firstPageText: "<<",
    nextPageText: ">",
    prePageText: "<",
    showTotal: true,
    paginationTotalRenderer: (from, to, size) => {
      return ` Exibindo registros ${from} até ${to} de ${size}`;
    },
    alwaysShowAllBtns: true,
  });

  const { SearchBar } = Search;

  return (
    <Container>
      <br></br>
      <Row>
        <Col md={3}>
          <Form.Group className="mb-3">
            <Form.Control
              type="text"
              placeholder="Informe o horário"
              value={horario}
              autoFocus
              maxLength="5"
              onChange={(event) => {
                setHorario(event.target.value);
              }}
              ref={(input) => {
                if (input !== null) {
                  horarioInput = input;
                }
              }}
            />
          </Form.Group>
        </Col>
        <Col md={1}>
          <Button
            variant="primary"
            onClick={() => {
              registrar();
            }}
          >
            Incluir
          </Button>
        </Col>
      </Row>
      <br></br>
      <Row>
        <Col md={12}>
          <div>
            <ToolkitProvider
              keyField="idHorario"
              data={lista}
              columns={colunas}
              search
              exportCSV={{
                fileName: "usuarios.csv",
              }}
            >
              {(props) => (
                <div>
                  <SearchBar
                    srText="Filtre os resultados"
                    placeholder="Informe o critério"
                    {...props.searchProps}
                  />
                  <br></br>
                  <BootstrapTable
                    classes="react-bootstrap-table"
                    keyField="idHorario"
                    striped={true}
                    hover={true}
                    defaultSorted={defaultSorted}
                    pagination={pagination}
                    {...props.baseProps}
                  />
                  <div className={["d-flex", "justify-content-end"].join(" ")}>
                    <BotaoExportar {...props.csvProps}></BotaoExportar>
                  </div>
                </div>
              )}
            </ToolkitProvider>
          </div>
        </Col>
      </Row>
    </Container>
  );
}

export default ParametrizacaoHorario;
