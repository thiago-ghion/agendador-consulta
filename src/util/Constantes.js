const URL_API = 'http://localhost:5000';

export { URL_API };