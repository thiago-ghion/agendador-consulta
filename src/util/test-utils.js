import React from "react";
import { render } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
// As a basic setup, import your same slice reducers
import loginReducer from "../features/loginSlice";
import profissionalReducer from "../features/profissionalSlice";
import consultaReducer from "../features/consultaSlice";
import pacienteReducer from "../features/pacienteSlice";
import horarioReducer from "../features/horarioSlice";
import usuarioReducer from "../features/usuarioSlice";
import loadingReducer from "../features/loadingSlice";
import mensagemgReducer from "../features/mensagemSlice";

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({
      reducer: {
        login: loginReducer,
        consulta: consultaReducer,
        profissional: profissionalReducer,
        paciente: pacienteReducer,
        horario: horarioReducer,
        usuario: usuarioReducer,
        loading: loadingReducer,
        mensagem: mensagemgReducer,
      },
      preloadedState,
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
