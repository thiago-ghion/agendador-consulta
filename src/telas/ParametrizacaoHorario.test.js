import React from "react";
import { rest } from "msw";
import { setupServer } from "msw/node";
import { renderWithProviders } from "../util/test-utils";

import ParametrizacaoHorario from "./ParametrizacaoHorario";
import { fireEvent, screen, waitFor } from "@testing-library/react";

let mockListaHorario = [];
let estadoRegistrar = 201;
let mockRegistrar = {};
let estadoAtivar = 200;
let mockAtivar = {};
let estadoDesativar = 200;
let mockDesativar = {};

export const handlers = [
  rest.get("http://localhost:5000/v1/horario/listar", (req, res, ctx) => {
    return res(ctx.json(mockListaHorario));
  }),
  rest.post(
    "http://localhost:5000/v1/horario/registrar/08:00",
    (req, res, ctx) => {
      return res(ctx.status(estadoRegistrar), ctx.json(mockRegistrar));
    }
  ),
  rest.post("http://localhost:5000/v1/horario/ativar/1", (req, res, ctx) => {
    return res(ctx.status(estadoAtivar), ctx.json(mockAtivar));
  }),
  rest.post("http://localhost:5000/v1/horario/desativar/1", (req, res, ctx) => {
    return res(ctx.status(estadoDesativar), ctx.json(mockDesativar));
  }),
];

const server = setupServer(...handlers);

// Enable API mocking before tests.
beforeAll(() => server.listen());

// Reset any runtime request handlers we may add during the tests.
afterEach(() => server.resetHandlers());

// Disable API mocking after the tests are done.
afterAll(() => server.close());

beforeEach(() => {
  mockListaHorario = [];
  estadoRegistrar = 201;
  mockRegistrar = {};
  estadoAtivar = 200;
  mockAtivar = {};
  estadoDesativar = 200;
  mockDesativar = {};
});

test("Renderização componente", async () => {
  expect(renderWithProviders(<ParametrizacaoHorario />)).toMatchSnapshot();
});

test("Registrar horário - Sucesso", async () => {
  renderWithProviders(<ParametrizacaoHorario />);
  const horario = await waitFor(() =>
    screen.findByPlaceholderText("Informe o horário")
  );

  fireEvent.change(horario, { target: { value: "08:00" } });

  const incluir = await waitFor(() =>
    screen.findByRole("button", { name: /Incluir/i })
  );
  fireEvent.click(incluir);
});

test("Registrar horário - Falha", async () => {
  estadoRegistrar = 400;
  renderWithProviders(<ParametrizacaoHorario />);
  const horario = await waitFor(() =>
    screen.findByPlaceholderText("Informe o horário")
  );

  fireEvent.change(horario, { target: { value: "08:00" } });

  const incluir = await waitFor(() =>
    screen.findByRole("button", { name: /Incluir/i })
  );
  fireEvent.click(incluir);
});

test("Ativar horário", async () => {
  mockListaHorario = [{ idHorario: 1, horario: "08:00", indicadorAtivo: "N" }];
  renderWithProviders(<ParametrizacaoHorario />);

  const ativar = await waitFor(() =>
    screen.findByRole("button", { name: /Ativar/i })
  );
  fireEvent.click(ativar);
});

test("Desativar horário", async () => {
  mockListaHorario = [{ idHorario: 1, horario: "08:00", indicadorAtivo: "S" }];
  renderWithProviders(<ParametrizacaoHorario />);

  const desativar = await waitFor(() =>
    screen.findByRole("button", { name: /Desativar/i })
  );
  fireEvent.click(desativar);
});
