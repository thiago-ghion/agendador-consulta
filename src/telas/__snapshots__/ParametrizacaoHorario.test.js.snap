// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renderização componente 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <br />
        <div
          class="row"
        >
          <div
            class="col-md-3"
          >
            <div
              class="mb-3"
            >
              <input
                class="form-control"
                maxlength="5"
                placeholder="Informe o horário"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="col-md-1"
          >
            <button
              class="btn btn-primary"
              type="button"
            >
              Incluir
            </button>
          </div>
        </div>
        <br />
        <div
          class="row"
        >
          <div
            class="col-md-12"
          >
            <div>
              <div>
                <label
                  class="search-label"
                  for="search-bar-0"
                >
                  <span
                    class="sr-only"
                    id="search-bar-0-label"
                  >
                    Filtre os resultados
                  </span>
                  <input
                    aria-labelledby="search-bar-0-label"
                    class="form-control "
                    id="search-bar-0"
                    placeholder="Informe o critério"
                    type="text"
                    value=""
                  />
                </label>
                <br />
                <div
                  class="react-bootstrap-table"
                >
                  <table
                    class="table table-striped table-hover table-bordered react-bootstrap-table"
                  >
                    <thead>
                      <tr>
                        <th
                          aria-label="# sort asc"
                          class="sortable"
                          style="width: 80px;"
                          tabindex="0"
                        >
                          #
                          <span
                            class="react-bootstrap-table-sort-order dropup"
                          >
                            <span
                              class="caret"
                            />
                          </span>
                        </th>
                        <th
                          aria-label="Horário sortable"
                          class="sortable"
                          tabindex="0"
                        >
                          Horário
                          <span
                            class="order"
                          >
                            <span
                              class="dropdown"
                            >
                              <span
                                class="caret"
                              />
                            </span>
                            <span
                              class="dropup"
                            >
                              <span
                                class="caret"
                              />
                            </span>
                          </span>
                        </th>
                        <th
                          aria-label="Ativação sortable"
                          class="sortable"
                          tabindex="0"
                        >
                          Ativação
                          <span
                            class="order"
                          >
                            <span
                              class="dropdown"
                            >
                              <span
                                class="caret"
                              />
                            </span>
                            <span
                              class="dropup"
                            >
                              <span
                                class="caret"
                              />
                            </span>
                          </span>
                        </th>
                      </tr>
                    </thead>
                  </table>
                </div>
                <div
                  class="row react-bootstrap-table-pagination"
                >
                  <div
                    class="col-md-6 col-xs-6 col-sm-6 col-lg-6"
                  >
                    <span
                      class="react-bs-table-sizePerPage-dropdown dropdown"
                      style="visibility: visible;"
                    >
                      <button
                        aria-expanded="false"
                        class="btn btn-default btn-secondary dropdown-toggle"
                        data-toggle="dropdown"
                        id="pageDropDown"
                        type="button"
                      >
                        10
                         
                        <span>
                          <span
                            class="caret"
                          />
                        </span>
                      </button>
                      <ul
                        aria-labelledby="pageDropDown"
                        class="dropdown-menu "
                        role="menu"
                      >
                        <li
                          class="dropdown-item"
                          role="presentation"
                        >
                          <a
                            data-page="10"
                            href="#"
                            role="menuitem"
                            tabindex="-1"
                          >
                            10
                          </a>
                        </li>
                        <li
                          class="dropdown-item"
                          role="presentation"
                        >
                          <a
                            data-page="25"
                            href="#"
                            role="menuitem"
                            tabindex="-1"
                          >
                            25
                          </a>
                        </li>
                        <li
                          class="dropdown-item"
                          role="presentation"
                        >
                          <a
                            data-page="30"
                            href="#"
                            role="menuitem"
                            tabindex="-1"
                          >
                            30
                          </a>
                        </li>
                        <li
                          class="dropdown-item"
                          role="presentation"
                        >
                          <a
                            data-page="50"
                            href="#"
                            role="menuitem"
                            tabindex="-1"
                          >
                            50
                          </a>
                        </li>
                      </ul>
                    </span>
                     Exibindo registros 0 até 0 de 0
                  </div>
                  <div
                    class="react-bootstrap-table-pagination-list col-md-6 col-xs-6 col-sm-6 col-lg-6"
                  >
                    <ul
                      class="pagination react-bootstrap-table-page-btns-ul"
                    />
                  </div>
                </div>
                <div
                  class="d-flex justify-content-end"
                >
                  <div>
                    <button
                      class="btn btn-primary"
                      type="button"
                    >
                      Exportar para CSV
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <br />
      <div
        class="row"
      >
        <div
          class="col-md-3"
        >
          <div
            class="mb-3"
          >
            <input
              class="form-control"
              maxlength="5"
              placeholder="Informe o horário"
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          class="col-md-1"
        >
          <button
            class="btn btn-primary"
            type="button"
          >
            Incluir
          </button>
        </div>
      </div>
      <br />
      <div
        class="row"
      >
        <div
          class="col-md-12"
        >
          <div>
            <div>
              <label
                class="search-label"
                for="search-bar-0"
              >
                <span
                  class="sr-only"
                  id="search-bar-0-label"
                >
                  Filtre os resultados
                </span>
                <input
                  aria-labelledby="search-bar-0-label"
                  class="form-control "
                  id="search-bar-0"
                  placeholder="Informe o critério"
                  type="text"
                  value=""
                />
              </label>
              <br />
              <div
                class="react-bootstrap-table"
              >
                <table
                  class="table table-striped table-hover table-bordered react-bootstrap-table"
                >
                  <thead>
                    <tr>
                      <th
                        aria-label="# sort asc"
                        class="sortable"
                        style="width: 80px;"
                        tabindex="0"
                      >
                        #
                        <span
                          class="react-bootstrap-table-sort-order dropup"
                        >
                          <span
                            class="caret"
                          />
                        </span>
                      </th>
                      <th
                        aria-label="Horário sortable"
                        class="sortable"
                        tabindex="0"
                      >
                        Horário
                        <span
                          class="order"
                        >
                          <span
                            class="dropdown"
                          >
                            <span
                              class="caret"
                            />
                          </span>
                          <span
                            class="dropup"
                          >
                            <span
                              class="caret"
                            />
                          </span>
                        </span>
                      </th>
                      <th
                        aria-label="Ativação sortable"
                        class="sortable"
                        tabindex="0"
                      >
                        Ativação
                        <span
                          class="order"
                        >
                          <span
                            class="dropdown"
                          >
                            <span
                              class="caret"
                            />
                          </span>
                          <span
                            class="dropup"
                          >
                            <span
                              class="caret"
                            />
                          </span>
                        </span>
                      </th>
                    </tr>
                  </thead>
                </table>
              </div>
              <div
                class="row react-bootstrap-table-pagination"
              >
                <div
                  class="col-md-6 col-xs-6 col-sm-6 col-lg-6"
                >
                  <span
                    class="react-bs-table-sizePerPage-dropdown dropdown"
                    style="visibility: visible;"
                  >
                    <button
                      aria-expanded="false"
                      class="btn btn-default btn-secondary dropdown-toggle"
                      data-toggle="dropdown"
                      id="pageDropDown"
                      type="button"
                    >
                      10
                       
                      <span>
                        <span
                          class="caret"
                        />
                      </span>
                    </button>
                    <ul
                      aria-labelledby="pageDropDown"
                      class="dropdown-menu "
                      role="menu"
                    >
                      <li
                        class="dropdown-item"
                        role="presentation"
                      >
                        <a
                          data-page="10"
                          href="#"
                          role="menuitem"
                          tabindex="-1"
                        >
                          10
                        </a>
                      </li>
                      <li
                        class="dropdown-item"
                        role="presentation"
                      >
                        <a
                          data-page="25"
                          href="#"
                          role="menuitem"
                          tabindex="-1"
                        >
                          25
                        </a>
                      </li>
                      <li
                        class="dropdown-item"
                        role="presentation"
                      >
                        <a
                          data-page="30"
                          href="#"
                          role="menuitem"
                          tabindex="-1"
                        >
                          30
                        </a>
                      </li>
                      <li
                        class="dropdown-item"
                        role="presentation"
                      >
                        <a
                          data-page="50"
                          href="#"
                          role="menuitem"
                          tabindex="-1"
                        >
                          50
                        </a>
                      </li>
                    </ul>
                  </span>
                   Exibindo registros 0 até 0 de 0
                </div>
                <div
                  class="react-bootstrap-table-pagination-list col-md-6 col-xs-6 col-sm-6 col-lg-6"
                >
                  <ul
                    class="pagination react-bootstrap-table-page-btns-ul"
                  />
                </div>
              </div>
              <div
                class="d-flex justify-content-end"
              >
                <div>
                  <button
                    class="btn btn-primary"
                    type="button"
                  >
                    Exportar para CSV
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
